modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/UpdatableTableImpl.java

```diff
@@ -383,35 +380,26 @@ private static void appendValue(RowAssembler rowAssembler, ColumnDescriptor colD
         rowAssembler.appendValue(val);
     }
 
-    private static <RowT> CompletableFuture<List<RowT>> handleInsertResults(
-            RowHandler<RowT> handler,
-            CompletableFuture<List<RowT>>[] futs
+    private <RowT> CompletableFuture<List<RowT>> handleInsertResults(
+            ExecutionContext<RowT> ectx,
+            Collection<RowBatch> batches
     ) {
-        return CompletableFuture.allOf(futs)
+        return collectRejectedRowsResponsesWithRestoreOrder(batches)
                 .thenApply(response -> {
-                    List<String> conflictRows = null;
-
-                    for (CompletableFuture<List<RowT>> future : futs) {
-                        List<RowT> values = future.join();
-
-                        if (nullOrEmpty(values)) {
-                            continue;
-                        }
-
-                        if (conflictRows == null) {
-                            conflictRows = new ArrayList<>(values.size());
-                        }
-
-                        for (RowT row : values) {
-                            conflictRows.add(handler.toString(row));
-                        }
+                    if (nullOrEmpty(response)) {
+                        return null;
                     }
 
-                    if (conflictRows != null) {
-                        throw conflictKeysException(conflictRows);
-                    }
+                    RowHandler<RowT> handler = ectx.rowHandler();
+                    IgniteTypeFactory typeFactory = ectx.getTypeFactory();
+                    RowSchema rowSchema = rowSchemaFromRelTypes(RelOptUtil.getFieldTypeList(desc.insertRowType(typeFactory)));
+                    RowHandler.RowFactory<RowT> rowFactory = handler.factory(rowSchema);
+
+                    List<String> conflictRows = response.stream()
```

Comment:
Please don't use stream.

Reply:
I catch what you mean, but it requires more code lines. It still is not required in our code style; let's leave it as is.

Reply:
It won't take so many lines of code, why not do it more efficiently right away?

Reply:
The same as before

Reply:
This piece of code is a candidate for the hot path, we need to fix it now.

Reply:
I desegree.